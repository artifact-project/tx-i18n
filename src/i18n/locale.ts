let defaultLocale: Locale = {};

export type Locale = {
	[phrase:string]: string;
}

export function setDefaultLocale(local: Locale) {
	defaultLocale = local;
}

export function getTranslate(phrase: string) {
	if (defaultLocale.hasOwnProperty(phrase)) {
		return defaultLocale[phrase];
	}

	// todo: not translated phrase
	return phrase;
}