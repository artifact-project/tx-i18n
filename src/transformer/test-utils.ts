import * as ts from 'typescript';
import { existsSync, readFileSync } from 'fs';
import i18nTxt from './transformer';

const compilerOptions: ts.CompilerOptions = {
    module: ts.ModuleKind.ESNext,
    target: ts.ScriptTarget.ESNext,
    jsx: ts.JsxEmit.Preserve,
    removeComments: false,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
}

export function transform(name: string, options: ts.CompilerOptions = {}) {
    const path = `${__dirname}/fixture/${name}`;
    const fileName = existsSync(`${path}.ts`) ? `${path}.ts` : `${path}.tsx`;
    const output = ts.transpileModule(readFileSync(fileName) + '', {
        fileName: fileName,
        compilerOptions: {
            ...compilerOptions,
            ...options
        },
        transformers: {
            before: [
                i18nTxt({
                    fnName: '__',
                    exclude: [],
                }),
            ],
        },
    });

    return `${output.outputText.trim()}\n`;
}
