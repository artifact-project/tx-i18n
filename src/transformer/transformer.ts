import * as ts from 'typescript';
import { R_ENTITIES, decodeEntities } from './entities';

type HumanTextChecker = (text: string, node: ts.Node) => boolean;

interface Config {
	isHumanText: HumanTextChecker;
	normalizeText: (value: string) => string;
	isTranslatableJsxAttribute: (attr: ts.JsxAttribute, element: ts.JsxOpeningLikeElement) => boolean;
	overrideHumanTextChecker?: (isHumanText: HumanTextChecker) => HumanTextChecker;
	fnName: string;
	packageName: string
	fileName: string;
	include: string[];
	exclude: string[];
	pharsesStore?: ContextedPhrases;
	compilerOptions: ts.CompilerOptions;
	imports: {
		[name:string]: ts.ImportDeclaration;
	};
	sourceFile: ts.SourceFile;
}

const R_IS_TAG = /^[a-z0-9:]+$/;

function log(obj: object, ind = '', max = 3) {
	if (obj == null || /number|string|boolean/.test(typeof obj)) {
		console.log(obj);
		return;
	}

	const copy = {...obj};
	let lines = [];
	delete copy['parent'];

	for (let key in copy) {
		lines.push(
			`${ind}${key}: ` + (
			copy[key] === null ? 'null' :
				typeof copy[key] === 'object'
					? max > 0 ? '\n' + log(copy[key], ind + '  ', max - 1) : '{ ... }'
					: copy[key]
		));
	}

	if (ind === '') {
		console.log(lines.join('\n'));
	}

	return lines.join('\n');
}

export type Pharse = {
	value: string;
	file: string;
	ctx: string;
	loc: {
		start: {
			line: number;
			character: number;
		};
		end: {
			line: number;
			character: number;
		};
	};
}

export type ContextedPhrases = {
	[context: string]: Pharse[];
}

export function contextedPhrasesForEach(
	phrases: ContextedPhrases,
	iterator: (phrase: Pharse, ctx: string) => void,
) {
	Object.keys(phrases).forEach((ctx) => {
		phrases[ctx].forEach((phrase) => {
			iterator(phrase, ctx);
		});
	})
}

export function contextedPhrasesFilter(
	phrases: ContextedPhrases,
	filter: (phrase: Pharse, ctx: string) => boolean,
) {
	const filteredPhrases = {} as ContextedPhrases;

	contextedPhrasesForEach(phrases, (phrase, ctx) => {
		if (filter(phrase, ctx)) {
			if (!filteredPhrases.hasOwnProperty(ctx)) {
				filteredPhrases[ctx] = [];
			}

			filteredPhrases[ctx].push(phrase);
		}
	});

	return filteredPhrases;
}

let phrasesContext: string[];
let phrasesWithContext: ContextedPhrases = {};
let useNormalizeText: boolean = true; // @todo: пока не работает

function normalizeText({normalizeText}: Config, value: string): string {
	return useNormalizeText ? normalizeText(value) : value;
}

function savePhrase(value: string, node: ts.Node, {sourceFile}: Config) {
	const context = phrasesContext[0];

	if (!phrasesWithContext.hasOwnProperty(context)) {
		phrasesWithContext[context] = [];
	}

	phrasesWithContext[context].push({
		ctx: context,
		value,
		file: sourceFile.fileName,
		loc: {
			start: sourceFile.getLineAndCharacterOfPosition(node.getStart()),
			end: sourceFile.getLineAndCharacterOfPosition(node.getEnd()),
		},
	});
}

export function resetPhrases() {
	Object.keys(phrasesWithContext).forEach(key => {
		delete phrasesWithContext[key];
	});
}

export function getPhrases() {
	return phrasesWithContext;
}

function createImport(name, path) {
	return ts.createImportDeclaration(
		undefined,
		undefined,
		ts.createImportClause(ts.createIdentifier(name), undefined),
		ts.createLiteral(path)
	);
}

function addImport(imports: Config['imports'], name, path) {
	// imports[name] = createImport(name, path);
	// TypeError: Cannot set property text of #<IdentifierObject> which has only a getter

	if (!imports.hasOwnProperty(name)) {
		Object.defineProperty(imports, name, {
			enumerable: true,
			value: createImport(name, path),
		});
	}

	return imports[name];
}

function useNativeImport(module: ts.ModuleKind) {
	return module === ts.ModuleKind.ESNext || module === ts.ModuleKind.ES2015;
}

function i18nId(cfg: Config, prop?: string) {
	const importDecl = addImport(cfg.imports, cfg.fnName, cfg.packageName);
	let id: ts.Expression = useNativeImport(cfg.compilerOptions.module)
		? ts.createIdentifier(cfg.fnName)
		: ts.getGeneratedNameForNode(importDecl);

	if (!useNativeImport(cfg.compilerOptions.module)) {
		id = ts.createPropertyAccess(id, 'default');
	}

	return prop ? ts.createPropertyAccess(id, prop) : id;
}

function i18nWrap(cfg: Config, prop: string, args: ts.Expression[]) {
	if (phrasesContext.length > 1) {
		if (args.length === 1) {
			args.push(ts.createArrayLiteral());
		}

		args.push(createLiteral(phrasesContext[0]));
	}

	return ts.createCall(i18nId(cfg, prop), [], args);
}

function visited<T extends ts.Node>(node: T): T {
	node['__visited__'] = true;
	return node;
}

function createLiteral(val: string) {
	const node = ts.createIdentifier(JSON.stringify(val));
	return visited(node);
}

function hasJsxTextChildren(node: ts.JsxElement | ts.JsxFragment, cfg: Config) {
	return node.children.some(child => {
		if (ts.isJsxText(child)) {
			return cfg.isHumanText(child.getText(), child)
		} else {
			return false;
		}
	});
}

function wrapStringLiteral(node: ts.StringLiteralLike, cfg: Config, decode?: boolean) {
	let text = normalizeText(cfg, node.getText());

	if (!cfg.isHumanText(text, node)) {
		return visited(node);
	}

	savePhrase(text.slice(1, -1), node, cfg);

	if (decode) {
		text = decodeEntities(text);
	}

	return i18nWrap(cfg, null, [ts.createIdentifier(text)]);
}

function wrapTemplateExpression(node: ts.TemplateExpression, context, cfg: Config) {
	const args = [];
	let phrase = node.head.text;

	node.templateSpans.forEach((span, idx) => {
		phrase += `<#${(idx + 1)}>${span.literal.text}`;
		args.push(visitNodeAndChildren(span.expression, context, cfg));
	});

	if (!cfg.isHumanText(phrase, node)) {
		return visited(node);
	}

	phrase = normalizeText(cfg, phrase);
	savePhrase(phrase, node, cfg);

	return i18nWrap(cfg, null, [
		createLiteral(phrase),
		ts.createArrayLiteral(args),
	]);
}

function visitNode(node: ts.Node, context, cfg: Config): ts.Node {
	const {
		isHumanText,
	} = cfg;

	if (node['__visited__']) {
		return node;
	}

	if (
		(!node.parent || !ts.isImportDeclaration(node.parent))
		&& (
			ts.isStringLiteral(node)
			|| ts.isNoSubstitutionTemplateLiteral(node)
		)
		&& isHumanText(node.getFullText(), node)
	) {
		const {parent} = node;

		if (parent) {
			if (
				ts.isPropertyAssignment(parent) && parent.name === node
				|| ts.isElementAccessExpression(parent)
			) {
				return node;
			}


			const attr = ts.isJsxAttribute(parent)
				? parent
				: parent.parent && ts.isJsxAttribute(parent.parent) && parent.parent
			;

			// JSX Attribute
			if (attr) {
				if (
					!isTagName(attr.parent.parent.tagName.getText())
					|| cfg.isTranslatableJsxAttribute(attr, attr.parent.parent)
				) {
					const newNode = wrapStringLiteral(node, cfg, true);
					return newNode === node ? node : ts.createJsxExpression(undefined, newNode);
				}
				return node;
			}
		}

		return wrapStringLiteral(node, cfg);
	} else if (ts.isTemplateExpression(node)) {
		const {parent} = node;

		if (parent && (ts.isComputedPropertyName(parent) || ts.isElementAccessExpression(parent))) {
			return node;
		}

		if (
			node.parent
			&& node.parent.parent
			&& ts.isJsxAttribute(node.parent.parent)
			&& (
				!cfg.isTranslatableJsxAttribute(node.parent.parent, node.parent.parent.parent.parent)
				&& isTagName(node.parent.parent.parent.parent.tagName.getText())
			)
		) {
			return node;
		}

		return wrapTemplateExpression(node, context, cfg);
	} else if ((ts.isJsxElement(node) || ts.isJsxFragment(node)) && hasJsxTextChildren(node, cfg)) {
		const args = [];
		let simple = true;
		let gpart = 0;
		let phrase = '';
		let hasText = false;

		node.children.forEach(function processing(child) {
			if (ts.isJsxElement(child)) {
				let part = ++gpart;

				simple = false;
				phrase += `<${part}>`;
				args.push(jsxTagToObject(child, context, cfg));
				child.children.forEach(processing)
				phrase += `</${part}>`;
			} else if (ts.isJsxText(child)) {
				hasText = true;
				phrase += child.getFullText().replace(/[\n\t]/g, '');
			} else if (ts.isJsxExpression(child)) {
				if (child.expression) {
					simple = false;
					phrase += `<#${++gpart}>`;
					args.push(visitNode(child.expression, context, cfg));
				}
			} else if (ts.isJsxSelfClosingElement(child)) {
				simple = false;
				phrase += `<${++gpart}/>`;
				args.push(jsxTagToObject(child, context, cfg));
			} else {
				log(child);
				throw new Error('Not supproted');
			}
		}, args);

		phrase = normalizeText(cfg, phrase.trim());

		if (!hasText || !isHumanText(phrase, node)) {
			return node;
		}

		savePhrase(phrase, node, cfg);

		if (simple && !R_ENTITIES.test(phrase)) {
			if (ts.isJsxFragment(node)) {
				return ts.updateJsxFragment(
					node,
					node.openingFragment,
					[ts.createJsxExpression(undefined, i18nWrap(
						cfg,
						null,
						[].concat(
							createLiteral(phrase),
							args.length ? ts.createArrayLiteral(args) : [],
						),
					))],
					node.closingFragment,
				);
			} else {
				return ts.updateJsxElement(
					node,
					node.openingElement,
					[ts.createJsxExpression(undefined, i18nWrap(
						cfg,
						null,
						[].concat(
							createLiteral(phrase),
							args.length ? ts.createArrayLiteral(args) : [],
						),
					))],
					node.closingElement,
				);
			}
		}

		const callExp = i18nWrap(
			cfg,
			'jsx',
			[
				createLiteral(phrase),
				ts.createArrayLiteral([
					(ts.isJsxFragment(node)
						? ts.createIdentifier('{ type: React.Fragment }')
						: jsxTagToObject(node, context, cfg)
					),
					...args,
				]),
			],
		);

		if (ts.isJsxElement(node.parent)) {
			return ts.createJsxExpression(undefined, callExp);
		}

		return callExp;
	}

	return node;
}

type InlineConfig = {
	context: string;
	normalize: string;
	isJSXExpr: boolean;
};

function parseInlineConfig(node: ts.Node): InlineConfig {
	let expr = '';

	if (ts.isJsxExpression(node) && !node.expression) {
		try {
			expr = node.getText();
		} catch (err) {}
	}

	if (node.hasOwnProperty('jsDoc') && node['jsDoc'].length) {
		expr = node['jsDoc'][0].getText();
	}

	const rule = expr ? expr.match(/@tx-i18n\s+([a-z]+(?:\s*:\s*[\w\d]+)?(?:,\s*)?)+/) : null;

	if (rule) {
		return rule[0]
			.split('@tx-i18n')[1]
			.trim()
			.split(/\s*,\s*/g).reduce((cfg, token) => {
				let [key, val] = token.split(/\s*:\s*/);
				cfg[key] = val == null ? 'true' : val;
				return cfg;
			}, {
				isJSXExpr: ts.isJsxExpression(node),
			} as InlineConfig)
		;
	}
}

function visitNodeAndChildren(node: ts.Node, context, cfg: Config, jsxInlineConfig: InlineConfig[] = []) {
	return ts.visitEachChild(
		visitNode(node, context, cfg),
		(childNode) => {
			let inlineCfg = parseInlineConfig(childNode);

			if (!inlineCfg) {
				if (ts.isJsxElement(childNode)) {
					inlineCfg = jsxInlineConfig.pop();
				}
			} else if (inlineCfg.isJSXExpr) {
				jsxInlineConfig.push(inlineCfg);
				inlineCfg = null;
			}

			if (inlineCfg) {
				useNormalizeText = inlineCfg.normalize !== 'false';
				inlineCfg.context && phrasesContext.unshift(inlineCfg.context);
			} else {
				useNormalizeText = true;
			}

			const result = visitNodeAndChildren(childNode, context, cfg);

			if (inlineCfg) {
				inlineCfg.context && phrasesContext.shift();
			}

			return result;
		},
		context,
	);
}

function isTagName(name: string) {
	return R_IS_TAG.test(name);
}

function jsxTagToObject(
	node: ts.JsxElement | ts.JsxSelfClosingElement,
	context: object,
	cfg: Config,
) {
	const element = ts.isJsxElement(node) ? node.openingElement : node;
	const {attributes} = element;

	let type = element.tagName.getFullText();
	let props: ts.ObjectLiteralElementLike[] = [];

	attributes.properties.forEach(prop => {
		if (ts.isJsxAttribute(prop)) {
			const {initializer} = prop;
			let name = prop.name.getText();
			let value: ts.Expression;

			if (initializer) {
				if (cfg.isTranslatableJsxAttribute(prop, element) && cfg.isHumanText(prop.getText(), node)) {
					if (ts.isJsxExpression(initializer)) {
						if (ts.isTemplateExpression(initializer.expression)) {
							value = wrapTemplateExpression(initializer.expression, context, cfg);
						} else {
							value = ts.createIdentifier(initializer.getText().slice(1, -1));
						}
					} else {
						value = wrapStringLiteral(initializer, cfg);
					}
				} else {
					value = ts.isJsxExpression(initializer) ? initializer.expression : initializer;
				}
			} else {
				value = ts.createIdentifier('true');
			}

			props.push(ts.createPropertyAssignment(
				stringifyObjectKey(name),
				value,
			));
		} else if (ts.isJsxSpreadAttribute(prop)) {
			// log();
			props.push(ts.createSpreadAssignment(prop.expression));
		}
	});


	return ts.createObjectLiteral([
		ts.createPropertyAssignment(
			'type',
			isTagName(type) ? ts.createIdentifier(`"${type}"`) : element.tagName,
		),
		ts.createPropertyAssignment(
			'props',
			props.length ? ts.createObjectLiteral(props) : ts.createNull(),
		),
	]);
}

function checkPatterFile(this: Config, mask: string) {
	return this.fileName.match(mask) !== null;
}

const R_IS_SIMPLE_OBJ_KEY = /^[a-z\d_]+$/i;
function stringifyObjectKey(name: string) {
	return R_IS_SIMPLE_OBJ_KEY.test(name) ? name : JSON.stringify(name)
}

export type TXConfig = Partial<Pick<Config,
	'fnName'
	| 'packageName'
	| 'include'
	| 'exclude'
	| 'isHumanText'
	| 'isTranslatableJsxAttribute'
	| 'overrideHumanTextChecker'
	| 'normalizeText'
	| 'pharsesStore'
>>

export default function transformerFactory(config: TXConfig) {
	if (config.pharsesStore) {
		phrasesWithContext = config.pharsesStore;
	}

	return function transformer(context: ts.TransformationContext) {
		return function visitor(file: ts.SourceFile) {
			const cfg: Config = {
				fnName: '__',
				packageName: 'tx-i18n',
				fileName: file.fileName,
				exclude: ['/tx-i18n/', '/node_modules/'],
				include: null,
				imports: {},
				compilerOptions: context.getCompilerOptions(),
				normalizeText: (value: string) => value.replace(/\s+/g, ' '),
				isHumanText: (value: string) => /[\wа-яё]/i.test(
					value.trim()
						.replace(/<\d+\/?>/g, '')
						.replace(/\d+([^\s]+)?/g, '')
						.trim()
				),
				isTranslatableJsxAttribute: (attr: ts.JsxAttribute) => /^(title|alt|placeholder|value)$/.test(attr.name.getText()),
				sourceFile: file,
				...config,
			};

			phrasesContext = ['default'];

			if (cfg.overrideHumanTextChecker) {
				cfg.isHumanText = cfg.overrideHumanTextChecker(cfg.isHumanText);
			}

			if (cfg.exclude.some(checkPatterFile, cfg)) {
				return file;
			}

			if (cfg.include && !cfg.include.some(checkPatterFile, cfg)) {
				return file;
			}

			try {
				const result = visitNodeAndChildren(file, context, cfg);

				try {
					return ts.updateSourceFileNode(
						result,
						Object.keys(cfg.imports)
							.map(name => cfg.imports[name])
							.concat(result.statements),
					);
				} catch (err) {
					console.error(`\x1b[31m\n[tx-i18n] [update] ${cfg.fileName}\n---\n${err.toString()}\n\x1b[0m`);
				return file;
				}
			} catch (err) {
				console.error(`\x1b[31m\n[tx-i18n] [visit] ${cfg.fileName}\n---\n${err.toString()}\n\x1b[0m`);
				return file;
			}
		}
	}
}
