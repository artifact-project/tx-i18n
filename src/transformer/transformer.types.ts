import * as ts from 'typescript';

export type HumanTextChecker = (text: string, node: ts.Node) => boolean;

export interface Config {
	verbose: (...args: any[]) => void;
	isHumanText: HumanTextChecker;
	normalizeText: (value: string) => string;
	isTranslatableJsxAttribute: (attr: ts.JsxAttribute, element: ts.JsxOpeningLikeElement) => boolean;
	overrideHumanTextChecker?: (isHumanText: HumanTextChecker) => HumanTextChecker;
	textFnName: string;
	jsxFnName: string;
	packageName: string
	fileName: string;
	include: string[];
	exclude: string[];
	pharsesStore?: ContextedPhrases;
	compilerOptions: ts.CompilerOptions;
	imports: {
		[name:string]: ts.ImportDeclaration;
	};
	sourceFile: ts.SourceFile;
};

export type Phrase = {
	value: string;
	file: string;
	ctx: string;
	loc: {
		start: {
			line: number;
			character: number;
		};
		end: {
			line: number;
			character: number;
		};
	};
};

export type ContextedPhrases = {
	[context: string]: Phrase[];
};

export type InlineConfig = {
	context: string;
	normalize: string;
	isJSXExpr: boolean;
};

export type TXConfig = Partial<Pick<Config,
	| 'textFnName'
	| 'jsxFnName'
	| 'packageName'
	| 'include'
	| 'exclude'
	| 'isHumanText'
	| 'isTranslatableJsxAttribute'
	| 'overrideHumanTextChecker'
	| 'normalizeText'
	| 'pharsesStore'
	| 'verbose'
>>;
