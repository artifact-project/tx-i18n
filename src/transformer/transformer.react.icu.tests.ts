import { transform, reactLoad, reactRender } from './test-utils';
import { setDefaultLocale, setLocale, setLang } from '../i18n/locale';
import { getPhrases, resetPhrases } from './transformer';
import { plural } from '../../icu/plural/en';
import { plural as ruPlural } from '../../icu/plural/ru';


beforeEach(() => {
	setLang('default')
	resetPhrases();
	setDefaultLocale({}, plural as any);
})

describe('react-icu', () => {
	describe('transform', () => {
		it('Info', () => {
			expect(transform('react-icu')).toMatchSnapshot();
			expect(getPhrases().default.map(v => v.value)).toEqual([
				'You have <a1>{v2, plural, one {# unread message} other {# unread messages}}</a1>.',
				'Total:',
				'{v1, plural, one {# message} other {# messages}}',
			]);
		});
	});

	describe('render', () => {
		const blocks = reactLoad('react-icu');

		it('Info', () => {
			expect(reactRender(blocks.Info, {unread: 1, total: 13})).toMatchSnapshot();
		});

		it('Info: ru', () => {
			setLang('ru');
			setLocale('ru', {
				default: {
					'You have <a1>{v2, plural, one {# unread message} other {# unread messages}}</a1>.': `У вас <a1>{v2, plural,
						one {# непрочитанное сообщение}
						few {# непрочитанных сообщения}
						many {# непрочитанных сообщений}
						other {# непрочитанных сообщений}
					}</a1>.`,
					'Total:': 'Всего:',
					'{v1, plural, one {# message} other {# messages}}': `{v1, plural,
						one {# сообщение}
						few {# сообщения}
						many {# сообщений}
						other {# сообщений}
					}`,
				},
			}, ruPlural);
			expect(reactRender(blocks.Info, {unread: 1, total: 13})).toMatchSnapshot();
		});

		it('Info: ru (partial)', () => {
			setLang('ru');
			setLocale('ru', {
				default: {
					'{v1, plural, one {# message} other {# messages}}': `{v1, plural,
						one {# сообщение}
						few {# сообщения}
						many {# сообщений}
						other {# сообщений}
					}`,
				},
			}, ruPlural);
			expect(reactRender(blocks.Info, {unread: 1, total: 13})).toMatchSnapshot();
		});
	});
});