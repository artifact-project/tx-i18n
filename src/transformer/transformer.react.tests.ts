import * as ts from 'typescript';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { transform } from './test-utils';
import i18n from '../../index';
import { setDefaultLocale } from '../i18n/locale';
import { getPhrases } from './transformer';

beforeEach(() => {
	setDefaultLocale({});
})

describe('transform', () => {
	it('react', () => {
		expect(transform('react')).toMatchSnapshot();
	});

	it('react-composite', () => {
		expect(transform('react-composite')).toMatchSnapshot();
	});
});

describe('render', () => {
	function load(name): any {
		try {
			const exports = {};
			const source = transform(name, {
				module: ts.ModuleKind.CommonJS,
				target: ts.ScriptTarget.ES5,
				jsx: ts.JsxEmit.React,
			});

			Function('exports, require', source)(
				exports,
				(name) => {
					if (name === 'tx-i18n') {
						return {default: i18n};
					} else if (name.substr(0, 2) === './') {
						return load(name.substr(2));
					} else {
						return require(name);
					}
				},
			);

			return exports;
		} catch (err) {
			console.error(err);
		}
	}

	function render(Block, props = {}) {
		return renderer
			.create(React.createElement(Block, props))
			.toJSON()
		;
	}

	describe('react', () => {
		const blocks = load('react');

		it('react: Static', () => {
			expect(render(blocks.Static)).toMatchSnapshot();
		});

		it('react: Entities', () => {
			expect(render(blocks.Entities)).toMatchSnapshot();
		});

		it('react: Hello', () => {
			expect(render(blocks.Hello, {username: 'Рубаха'})).toMatchSnapshot();
		});

		it('react: ClickHere', () => {
			expect(render(blocks.ClickHere, {username: 'Рубаха'})).toMatchSnapshot();
		});

		it('react: Fragment', () => {
			expect(render(blocks.Fragment)).toMatchSnapshot();
		});
	});

	describe('react-composite', () => {
		const blocks = load('react-composite');

		it('react-composite: DeepHello', () => {
			setDefaultLocale({
				default: {
					'Привет <#1>-кун!1': 'Hi, <#1>-kun!',
					'<1/> или Нет!': '<1/> or No!',
				},
			});
			expect(render(blocks.DeepHello, {username: 'i18n'})).toMatchSnapshot();
		});

		it('react-composite: DeepHello (invert)', () => {
			setDefaultLocale({
				default: {
					'Привет <#1>-кун!1': '<#1>, HI!',
					'<1/> или Нет!': 'No or <1/>!',
				},
			});
			expect(render(blocks.DeepHello, {username: 'i18n'})).toMatchSnapshot();
		});

		it('react-composite: Dialog', () => {
			setDefaultLocale({
				default: {
					'Подвал': 'Footer',
					'Хорошо': 'OK',
					'Отмена': 'Cancel',
					'Ширина:': 'Width:',
					'Да<#1><2/>': '<2/> — fail',
					'Тест \"фу\" бар': 'Test "foo" bar',
				},
				form: {
					'Сохранить': 'Save',
					'<1/>или<2/>': '<1/> or <2/>',
					'Отмена': 'Form cancel',
					'Да<#1><2/>': '<2/> — YES',
				},
			});

			expect(Object.keys(getPhrases())).toEqual(['default', 'form']);
			expect(render(blocks.Dialog)).toMatchSnapshot();
		});
	});
});
