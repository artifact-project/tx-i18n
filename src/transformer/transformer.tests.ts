import { transform } from './test-utils';
import { getPhrases, resetPhrases } from './transformer';

beforeEach(() => {
	resetPhrases();
})

it('empty', () => {
	expect(transform('empty')).toMatchSnapshot(`'';`);
});

it('numbers', () => {
	expect(transform('numbers')).toMatchSnapshot();
});

it('simple', () => {
	expect(transform('simple')).toMatchSnapshot();
});

it('object', () => {
	expect(transform('object')).toMatchSnapshot();
});

it('template', () => {
	expect(transform('template')).toMatchSnapshot();
});

it('func', () => {
	expect(transform('func')).toMatchSnapshot();
	// expect(getPhrases().default[0].value).toEqual('Single "quote"')
	// expect(getPhrases().default[1].value).toEqual('Double \"quote\"');
	// expect(getPhrases().default[2].value).toEqual('Mixed \"quote\"');
});

it('phrases', () => {
	transform('simple')

	expect(Object.keys(getPhrases())).toEqual(['default']);
	expect(getPhrases().default[0]).toEqual({
		ctx: 'default',
		file: `${__dirname}/fixture/simple.ts`,
		value: 'foo',
		loc: {
			start: {line: 0, character: 0},
			end: {line: 0, character: 5},
		},
	});
});

it('phrases with context', () => {
	resetPhrases();
	transform('context')

	expect(Object.keys(getPhrases())).toEqual([
		'default',
		'other',
	]);

	expect(getPhrases().default.map(p => p.value)).toEqual([
		'foo',
		'bar',
		'qux',
	]);

	expect(getPhrases().other[0].ctx).toEqual('other');
	expect(getPhrases().other.map(p => p.value)).toEqual([
		'bar',
		'foo',
		'bar',
		'baz',
		'qux',
	]);
});
