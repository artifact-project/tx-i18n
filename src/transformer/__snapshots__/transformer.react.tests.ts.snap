// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render react react: ClickHere 1`] = `
<div>
  Привет 
  Рубаха
  , нажми 
  <a
    href="#"
  >
    здесь
  </a>
   чтобы продолжить.
</div>
`;

exports[`render react react: Entities 1`] = `
<div
  className="host"
>
  <p
    title="Foo bar"
  >
    <h2>
      Left — Right
    </h2>
  </p>
  <div>
    <div
      className="input text"
    >
      <input
        placeholder="Подсказка: Space — bar"
      />
      <button
        value="Войти"
      />
    </div>
  </div>
  <div>
    Enter:
    <div
      className="input text"
    >
      <input
        placeholder="Подсказка: User‑name"
      />
      <button
        value="Войти"
      />
    </div>
  </div>
</div>
`;

exports[`render react react: Fragment 1`] = `
<div>
  Простой текст
  Фрагмент и 
  <i />
</div>
`;

exports[`render react react: Hello 1`] = `
<div
  className="host"
>
  Привет 
  Рубаха
  -кун!1
</div>
`;

exports[`render react react: Static 1`] = `
<div
  className="host"
>
  Вова, купи "гречу"!
</div>
`;

exports[`render react-composite react-composite: DeepHello (invert) 1`] = `
<div>
  No or 
  <div
    className="host"
  >
    i18n
    , HI!
  </div>
  !
</div>
`;

exports[`render react-composite react-composite: DeepHello 1`] = `
<div>
  <div
    className="host"
  >
    Hi, 
    i18n
    -kun!
  </div>
   or No!
</div>
`;

exports[`render react-composite react-composite: Dialog 1`] = `
<div
  className="host"
>
  <h1>
    Заголовок 
    <i />
     да её имультилайн, да-да.
  </h1>
  <h3>
    Test "foo" bar
  </h3>
  <form>
    <h2>
       Ох, а вот ещё один мультилайн заголовок. 
    </h2>
    <fieldset>
      <div
        className="row"
      >
        <div
          className="col"
        >
          Width:
        </div>
        <div
          className="col"
        >
          123px
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col"
        >
          <input
            checked={true}
            type="checked"
          />
           — YES
        </div>
      </div>
      <div>
        Окей
      </div>
      <p>
        <button
          value="Form cancel"
        />
         or 
        <button
          name="save"
          title={undefined}
          type="submit"
          value="Save"
        />
      </p>
    </fieldset>
    <div>
      <div
        title="Footer"
      >
        <button
          name="submit"
          title="Подсказка: 123!"
          type="button"
          value="OK"
        />
        <button
          data-id="рав-текст"
          value="Cancel"
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`transform react 1`] = `
"import __ from \\"tx-i18n\\";
import * as React from 'react';
export const Static = () => (__.jsx(\\"Вова, купи&nbsp;\\\\\\"гречу\\\\\\"!\\", [{ type: \\"div\\", props: { className: 'host' } }]));
export const Entities = () => (<div className={'host'}>
		<p title={__(\\"Foo bar\\")}>
			{__.jsx(\\"Left &mdash; Right\\", [{ type: \\"h2\\", props: null }])}
		</p>
		<div>
			<Input hint={__(\\"Space — bar\\")} mode={__(\\"text\\")}/>
		</div>
		{__.jsx(\\"Enter:<1/>\\", [{ type: \\"div\\", props: null }, { type: Input, props: { hint: (), mode: () } }])}
	</div>);
export const Hello = ({ username }) => (__.jsx(\\"Привет <#1>-кун!1\\", [{ type: \\"div\\", props: { className: \`host\` } }, username]));
export const ClickHere = ({ username }) => (__.jsx(\\"Привет&nbsp;<#1>, нажми <2>здесь</2> чтобы продолжить.\\", [{ type: \\"div\\", props: null }, username, { type: \\"a\\", props: { href: \\"#\\" } }]));
export const Input = ({ hint, mode }) => (<div className={\`input \${mode}\`}>
		
		<input placeholder={__(\\"Подсказка: <#1>\\", [hint])}/>
		<button value={__(\\"Войти\\")}/>
	</div>);
export const InputWithText = ({ hint, mode }) => (__.jsx(\\"Представьтесь: <1/><2/>\\", [{ type: \\"div\\", props: { className: \`input \${mode}\` } }, { type: \\"input\\", props: { placeholder: __(\\"Подсказка: <#1>\\", [hint]) } }, { type: \\"button\\", props: { value: __(\\"Войти\\") } }]));
export const Fragment = () => (<div>
		<>{__(\\"Простой текст\\")}</>
		{__.jsx(\\"Фрагмент и <1/>\\", [{ type: React.Fragment }, { type: \\"i\\", props: null }])}
	</div>);
"
`;

exports[`transform react-composite 1`] = `
"import __ from \\"tx-i18n\\";
import * as React from 'react';
import { Hello } from './react';
export const DeepHello = (props) => {
    return __.jsx(\\"<1/> или Нет!\\", [{ type: \\"div\\", props: null }, { type: Hello, props: { ...props } }]);
};
export const Button = ({ name, text, hint, primary }) => {
    return <button name={name} value={text} title={hint} type={primary ? __('submit') : __('button')}/>;
};
export const Form = (props) => {
    return (<form>
			{props.children}
			<div>
				{props.controls}
			</div>
		</form>);
};
export class Dialog extends React.Component {
    innerRender(props) {
    }
    render() {
        const { props } = this;
        const text = __(\\"Тест \\\\\\"фу\\\\\\" бар\\");
        return (<div className=\\"host\\">
				{__.jsx(\\"Заголовок <1/> да её имультилайн, да-да.\\", [{ type: \\"h1\\", props: null }, { type: \\"i\\", props: null }])}
				<h3>{text}</h3>

				<Form controls={<div title={__(\\"Подвал\\")}>
						<Button name={__(\\"submit\\")} text={__(\\"Хорошо\\")} hint={__(\\"Подсказка: <#1>!\\", [123])}/>
						<button value={__(\\"Отмена\\")} data-id={\`рав-текст\`}/>
					</div>}>
					<h2>{__(\` Ох, а вот ещё один мультилайн заголовок. \`)}</h2>
					
					<fieldset>
						<div className=\\"row\\">
							<div className=\\"col\\">{__(\\"Ширина:\\", [], \\"form\\")}</div>
							<div className=\\"col\\">123px</div>
						</div>
						<div className=\\"row\\">
							{__.jsx(\\"Да<#1><2/>\\", [{ type: \\"div\\", props: { className: \\"col\\" } }, ' ', { type: \\"input\\", props: { type: \\"checked\\", checked: true } }], \\"form\\")}
						</div>
						<div>
							{this.innerRender({ ...props, value: __('Тест', [], \\"form\\") })}
							Окей
						</div>
						{__.jsx(\\"<1/>или<2/>\\", [{ type: \\"p\\", props: null }, { type: \\"button\\", props: { value: __(\\"Отмена\\", [], \\"form\\") } }, { type: Button, props: { name: (), text: (), primary: true } }], \\"form\\")}
					</fieldset>
				</Form>
			</div>);
    }
}
"
`;
